Added parallel collation for fms and fixed a small bug in '-d' implementation

aidanheerdegen opened the pull request:
Paul's collation jobs were taking a long time, he was saving at high temporal resolution and doing ensembles, and my high temporal resolution 0.1 deg sims were taking 12+ hours to collate, so this is a fix of sorts.

I think it would be better to submit multiple collate jobs, but that would have been a bigger change, so I settled for using multiprocessing to submit multiple jobs within one collate job.

This adds config option collate_ncpus, to specify the number of cpus to be requested for the collate PBS submission.

The drawback is that multiprocessor jobs can't be submitted to the copy queue on raijin, so normal or express must be specified. Care must be taken to also request enough memory for multiple collate jobs to run simultaneously.

There was a bug in the '-d' directory path implementation, that it didn't work at all when invoked in a directory that wasn't the laboratory, which sort of defeats the purpose. In any case a minimal config.yaml must be supplied so that payu knows the model type. If the job name can't be determined payu-collate falls back to the name the PBS job after the directory path being collated. I had an issue with a trailing slash on the PBS job name interacting poorly with PBS. I sprinkled a few os.path.normpath calls around, but in the end the only one that matters is in collate_cmd.py. I left the others in as I didn't think they could do any harm, and might be beneficial.

marshallward commented:
Thanks, this looks interesting. Do we need to do a new release ASAP?

aidanheerdegen commented:
I've installed this as payu/dev version under /projects/v45 so Paul can access it. I'm not sure there is anyone else who needs any of these features just yet. It might be a good idea to test it for a while first in that case.

marshallward commented:
ok.
I am doing multidim support for f90nml so I'd rather get that working
before an update.

On Wed, Jul 8, 2015 at 4:25 PM, Aidan Heerdegen <notifications@github.com>
wrote:

> I've installed this as payu/dev version under /projects/v45 so Paul can
> access it. I'm not sure there is anyone else who needs any of these
> features just yet. It might be a good idea to test it for a while first in
> that case.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/marshallward/payu/pull/40#issuecomment-119451759>.
>

aidanheerdegen commented:
Yup. Might sneak a few more features in too ... :)

aidanheerdegen commented:
I stuffed up and introduced a bug, so I've fixed that and added a couple of other things, but I realised that by default, as it is currently written, when running payu-collate it will run in parallel and use as many CPUs as are available.

This could be considered a feature or a bug, depending on who pays your salary. 

Should I change this behaviour?

marshallward commented:
"as many CPUs as are available" as in as many requested for the PBS job? As many on the node? Or raijin?

I don't see the first as a problem. If you request 16 then you should use 16. The other cases could be a problem though.

aidanheerdegen commented:
Yeah, I realised that payu-run is still faithful to the ncpus setting in config.yaml, so payu-collate should be too. I'll change that before I commit & pull.

marshallward commented:
"ncpus" might be the problem.. maybe a collate_ncpus...
(or maybe a new collate subcategory)

On Thu, Jul 9, 2015 at 2:05 PM, Aidan Heerdegen <notifications@github.com>
wrote:

> Yeah, I realised that payu-run is still faithful to the ncpus setting in
> config.yaml, so payu-collate should be too. I'll change that before I
> commit & pull.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/marshallward/payu/pull/40#issuecomment-119809003>.
>

aidanheerdegen commented:
There is already a collate_ncpus config option, I just wasn't doing anything with it when payu-collate was called. I will honour that setting. As I said, makes it consistent, which is a virtue.

aidanheerdegen commented:
I don't think we can have a new collate subcategory without break previous scripts can we?

marshallward commented:
ah.. well you just check for both for backwards compatibility.. but its not
really necessary. Dont worry about it

On Thu, Jul 9, 2015 at 2:47 PM, Aidan Heerdegen <notifications@github.com>
wrote:

> I don't think we can have a new collate subcategory without break previous
> scripts can we?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/marshallward/payu/pull/40#issuecomment-119817124>.
>
